### PR TITLE
Add 64-bit LoongArch support

### PR DESCRIPTION
**Pull request by xen0n**
This is based on the WIP Linux port still under review, but the syscall
interface is remarkably similar to that of RISC-V, namely using
asm-generic numbers and parameters, so breakage should be minimal, if
at all, when the port is eventually merged.

Currently only the LP64 ABI is implemented in the toolchains, so only
support the host type "loongarch64".

* NEWS: Mention this.
* configure.ac [$host_cpu == loongarch64]: Define LOONGARCH64.
* src/linux/loongarch64/arch_regs.c: New file.
* src/linux/loongarch64/get_error.c: Likewise.
* src/linux/loongarch64/get_scno.c: Likewise.
* src/linux/loongarch64/get_syscall_args.c: Likewise.
* src/linux/loongarch64/ioctls_arch0.h: Likewise.
* src/linux/loongarch64/ioctls_inc0.h: Likewise.
* src/linux/loongarch64/raw_syscall.h: Likewise.
* src/linux/loongarch64/set_error.c: Likewise.
* src/linux/loongarch64/set_scno.c: Likewise.
* src/linux/loongarch64/syscallent.h: Likewise.
* src/Makefile.am (EXTRA_DIST): Add them.
* src/xlat/elf_em.in: Add EM_LOONGARCH.

### CODE REVIEW

**Comment by xen0n** [comment]
@loongarch64/dev-team Please review, thanks.

**Comment by cnmushiba** [comment]
make check发现下列三项测试没有过，前两个感觉还是得看一下原因
FAIL: pwritev.gen.test
FAIL: rt_sigreturn.gen.test
FAIL: pc.test

File: src/linux/loongarch64/get_error.c [hunk 1]
```diff
@@ -0,0 +1,19 @@
+/*
+ * Copyright (c) 2021 The strace developers.
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier: LGPL-2.1-or-later
+ */
+
+#include "negated_errno.h"
+
+static void
```
Comment (cnmushiba) [inline src/linux/loongarch64/get_error.c:10]: 刚开始看，这儿的static感觉有问题，后来看代码，这些C文件都是被include，然后没有疑问了。

Comment (xen0n) [reply src/linux/loongarch64/get_error.c:10]: 哈哈哈这些都是抄的 ;-)

**Comment by xen0n** [comment]
> make check发现下列三项测试没有过，前两个感觉还是得看一下原因
> FAIL: pwritev.gen.test
> FAIL: rt_sigreturn.gen.test
> FAIL: pc.test

晚上回家看看，最快这两周研究完再推一版

**Comment by yetist** [comment]
貌似上游有bug，
https://github.com/loongarch64/strace/blob/09740b8109f6fd9460dbe2fc16c652707bf24994/src/xlat/nl_xfrm_types.in#L34 应该参照内核定义为0x28，而不是0x26。
见内核代码：
https://github.com/torvalds/linux/blob/f9e36107ec70445fbdc2562ba5b60c0a7ed57c20/include/uapi/linux/xfrm.h#L222

**Comment by wdc596933938** [comment]
有几个测试跑不过pc.test/pwritev.test/

**Comment by wdc596933938** [comment]
FAIL: rt_sigreturn.gen
======================

+ ../strace -V
+ TIMEOUT=timeout -k 5 -s XCPU 600
+ timeout -k 5 -s XCPU 600 true
+ [ 1 -eq 0 ]
+ exec timeout -k 5 -s XCPU 600 ../../tests/rt_sigreturn.gen.test
+ run_strace_match_diff -esignal=!USR1
+ args=-esignal=!USR1
+ [ -n -esignal=!USR1 -a -z -esignal=!USR1 ]
+ set -- -e trace=rt_sigreturn -esignal=!USR1
+ run_prog
+ [ 0 -eq 0 ]
+ set -- ../rt_sigreturn
+ args=../rt_sigreturn
+ ../rt_sigreturn
+ run_strace -e trace=rt_sigreturn -esignal=!USR1 ../rt_sigreturn
+ 
+ args=-e trace=rt_sigreturn -esignal=!USR1 ../rt_sigreturn
+ ../../strace -o log -e trace=rt_sigreturn -esignal=!USR1 ../rt_sigreturn
+ match_diff log exp
+ local output expected error
+ [ 2 -eq 0 ]
+ output=log
+ shift
+ [ 1 -eq 0 ]
+ expected=exp
+ shift
+ [ 0 -eq 0 ]
+ error=../../strace -e trace=rt_sigreturn -esignal=!USR1 ../rt_sigreturn output mismatch
+ check_prog diff
+ type diff
+ diff -u -- exp log
--- exp	2021-10-21 03:22:32.980000000 +0000
+++ log	2021-10-21 03:22:32.980000000 +0000
@@ -1,2 +1,2 @@
-rt_sigreturn({mask=[INT USR2 CHLD RT_3 RT_4 RT_5 RT_26 RT_27]}) = 0
+rt_sigreturn({mask=[HUP INT QUIT ILL TRAP ABRT BUS FPE KILL USR1 SEGV USR2 PIPE ALRM TERM STKFLT CHLD CONT STOP TSTP TTIN TTOU URG XCPU XFSZ VTALRM PROF WINCH IO PWR SYS RTMIN RT_1 RT_2 RT_3 RT_4 RT_5 RT_6 RT_7 RT_8 RT_9 RT_10 RT_11 RT_12 RT_13 RT_14 RT_15 RT_16 RT_17 RT_18 RT_19 RT_20 RT_21 RT_22 RT_23 RT_24 RT_25 RT_26 RT_27 RT_28 RT_29 RT_30 RT_31 RT_32 RT_34 RT_36]}) = 0
 +++ exited with 0 +++
+ fail_ ../../strace -e trace=rt_sigreturn -esignal=!USR1 ../rt_sigreturn output mismatch
+ warn_ rt_sigreturn.gen.test: failed test: ../../strace -e trace=rt_sigreturn -esignal=!USR1 ../rt_sigreturn output mismatch
+ printf %s\n rt_sigreturn.gen.test: failed test: ../../strace -e trace=rt_sigreturn -esignal=!USR1 ../rt_sigreturn output mismatch
rt_sigreturn.gen.test: failed test: ../../strace -e trace=rt_sigreturn -esignal=!USR1 ../rt_sigreturn output mismatch
+ exit 1
FAIL rt_sigreturn.gen.test (exit status: 1)

**Comment by liyuzhao** [comment]
la64上有能用的strace了吗？

**Comment by xen0n** [comment]
> la64上有能用的strace了吗？

我可能近期空了 rebase 一下这个，然后也修下测试，这个 PR 目前的代码状态我一直在用，能用的

**Comment by xen0n** [comment]
This is now done in strace/strace#205, we're upstream!

**Pull request merged**